### ISSUE TITLE
Need Sidebar with sub heading menu 

### ISSUE THREAD

**Issue by ramko93** [kiwix/kiwix-js#1212]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

Kiwix extension chrome rendering zim article without subtopic menu and hyperlink for the same

**Comment by Greeshmanth1909** [comment]
@Jaifroid Would filtering the headings `<h1>, <h2> ...<h6>` in the `htmlArticle` parameter of `displayArticleContentInIframe` function and displaying them as a list, like `<a href="#idOfHeading">Heading</a>` at the bottom if the iframe work?

**Comment by Jaifroid** [comment]
@Greeshmanth1909 Yes, that's more or less the solution I made for the PWA. Did you look at the implementation? Go to https://pwa.kiwix.org, open a Wikipedia ZIM (can be a small one, like Ray Charles), and see how it works.

I think the code there is a bit clunky, not something I'm proud of, but it gives an idea of what could be done. The code is split between `uiUtil.js` ([here](https://github.com/kiwix/kiwix-js-pwa/blob/main/www/js/lib/uiUtil.js#L312)) and `app.js` (https://github.com/kiwix/kiwix-js-pwa/blob/main/www/js/app.js#L6842here). You have to set it up once the article is loaded, and then you have to add event listeners to each heading.

**Comment by Greeshmanth1909** [comment]
Nice, I'd like to take up this issue. 
Wouldn't it be better to put the TOCat the side to increase its visibility?

**Comment by Jaifroid** [comment]
> Nice, I'd like to take up this issue. Wouldn't it be better to put the TOC at the side to increase its visibility?

Well, we currently only have two UI options, top and bottom bars. We don't have a sidebar. It would be a much more complex thing to implement taking into account how to activate it, how to integrate it with the current UI, etc.

So, I think it would be best to implement it like in Kiwix PWA: it will be much easier for you to prototype and integrate.

**Comment by D3V-D** [comment]
@Jaifroid Is this issue still relevant? I may be interested in picking it up

**Comment by Jaifroid** [comment]
@D3V-D Apologies for the slow reply. The issue is indeed still relevant. In fact, a user just emailed me to ask:

> I can't see any way to find a list of the sections in the current
> wikipedia page.  How is that supposed to work?  Is there a way to move
> forward or back to the next section?

Your suggested solution is interesting, but it does appear to involve altering the HTML of the articles by injecting the ToC, and it might not be possible to do that in a generic way for all ZIMs with `<h2>`-style headers (not just Wikipedia). Another solution was that adopted in #1241, which was unfortunately abandoned very close the finishing line. That involved adding a "ToC" button (or other icon) to the current app UI, which is the solution also adopted and available in the PWA (https://pwa.kiwix.org).

![Image](https://github.com/user-attachments/assets/81d71536-9776-4b7b-af2a-f829641e4571)

I'm open to suggestions re the UI. If you wanted to work on this, it would be very welcome.

**Comment by D3V-D** [comment]
@Jaifroid would it not be possible to have the ToC as a floating HTML element above the current article, not necessarily _in_ the article? Though I do see the argument for a more out-of-the-way implementation. 

If I'm being honest, the bottom bar isn't my favorite UI/UX wise - if I were to choose, I'd probably emulate how modern browsers look, with the back/forward arrows in the top left corner, zoom in the top right, either side of search bar.  Back-to-top button perhaps is the only element I'd keep in the bottom. I do get why though, you already have a decent number of elements in the top-bar due to the different Kiwix app pages.

Past that, though, I think it'd be interesting to have ToC in top bar, like top-left, and have a sidebar pop-out (on a higher z than the main page, like your example) with the contents, instead of coming from the bottom.

Let me know your thoughts

**Comment by heropj** [comment]
Hello @Jaifroid, I have reviewed the solution adopted in #1241, the implementation is nearly finished, just a few tweaks are needed to achieve a well-structured TOC similar to what we have in kiwix-pwa.
If no one else is working on this currently, I'm ready to take this up.

**Comment by Jaifroid** [comment]
> Hello [@Jaifroid](https://github.com/Jaifroid), I have reviewed the solution adopted in [#1241](https://github.com/kiwix/kiwix-js/pull/1241), the implementation is nearly finished, just a few tweaks are needed to achieve a well-structured TOC similar to what we have in kiwix-pwa. If no one else is working on this currently, I'm ready to take this up.

@heropj Thanks for the offer. Yes, it was a shame that it got abandoned. If you'd like to complete it, please feel free to have a go.
